contracts/src/types/ProofRequest.sol

```diff
@@ -61,32 +62,36 @@ library ProofRequestLibrary {
         );
     }
 
-    /// @notice Verifies the client's signature over the proof request.
+    /// @notice Verifies a signature over a proof request.
+    /// @dev Supports EIP-1271 for smart contract signatures. If the signature is empty and the client is a smart
+    /// contract, calls isValidSignature with the abi encoded request as the signature.
     /// @param structHash The EIP-712 struct hash of the proof request.
     /// @param addr The address of the client.
     /// @param signature The signature to validate.
-    /// @return The struct hash if the signature is valid.
-    function verifyClientSignature(ProofRequest calldata, bytes32 structHash, address addr, bytes calldata signature)
+    function verifySignature(ProofRequest calldata request, bytes32 structHash, address addr, bytes calldata signature)
         internal
-        pure
-        returns (bytes32)
+        view
     {
-        if (ECDSA.recover(structHash, signature) != addr) {
-            revert IBoundlessMarket.InvalidSignature();
+        if (addr.code.length == 0) {
+            // Standard EOA flow.
+            if (ECDSA.recover(structHash, signature) != addr) {
+                revert IBoundlessMarket.InvalidSignature();
+            }
+        } else if (signature.length == 0) {
+            // If the signature is empty and the client is a smart contract, we provide the request as the signature.
+            // This flow is intended to enable anyone to submit requests on behalf of a client smart contract.
+            // The client smart contract is expected to validate the request was constructed correctly.
+            if (IERC1271(addr).isValidSignature(structHash, abi.encode(request)) != IERC1271.isValidSignature.selector)
+            {
+                revert IBoundlessMarket.InvalidSignature();
+            }
+        } else {
```

Comment:
Do we have a good idea of the value of this case? I would be inclined to remove it and always forward the passed signature data. In the case where the client does use `abi.encode(request)` as its EIP1271 signature, this would result in larger calldata. So this is really just an optimization for that case.

This is likely not a major issue, but this also introduces a bit of a completeness hole in that it is impossible to use an empty string as a signature, which might be relevant if e.g. the authorizing contract works by saving a set of validated request hashes.

Reply:
Removed